porcelain/storagedeals.go

```diff
@@ -12,20 +16,127 @@ var (
-	deals, err := plumbing.DealsLs()
+func DealGet(ctx context.Context, plumbing dealGetPlumbing, dealCid cid.Cid) (*storagedeal.Deal, error) {
+	dealCh, err := plumbing.DealsLs(ctx)
 	if err != nil {
 		return nil, err
 	}
-	for _, storageDeal := range deals {
-		if storageDeal.Response.ProposalCid == dealCid {
-			return storageDeal, nil
+	for deal := range dealCh {
+		if deal.Err != nil {
+			return nil, deal.Err
+		}
+		if deal.Deal.Response.ProposalCid == dealCid {
+			return &deal.Deal, nil
 		}
 	}
 	return nil, ErrDealNotFound
 }
+
+type dealClientLsPlumbing interface {
+	ConfigGet(string) (interface{}, error)
+	DealsLs(context.Context) (<-chan *StorageDealLsResult, error)
+}
+
+// DealClientLs returns a channel with all deals placed as a client
+func DealClientLs(ctx context.Context, plumbing dealClientLsPlumbing) (<-chan *StorageDealLsResult, error) {
+	minerAddress, _ := plumbing.ConfigGet("mining.minerAddress")
+
+	dealCh, err := plumbing.DealsLs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	outCh := filterDealChannel(dealCh, func(deal *storagedeal.Deal) bool {
+		return deal.Miner != minerAddress
+	})
+
+	return outCh, nil
+}
+
+type dealMinerLsPlumbing interface {
+	ConfigGet(string) (interface{}, error)
+	DealsLs(context.Context) (<-chan *StorageDealLsResult, error)
+}
+
+// DealMinerLs returns a channel with all deals received as a miner
+func DealMinerLs(ctx context.Context, plumbing dealMinerLsPlumbing) (<-chan *StorageDealLsResult, error) {
+	minerAddress, _ := plumbing.ConfigGet("mining.minerAddress")
+
+	dealCh, err := plumbing.DealsLs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	outCh := filterDealChannel(dealCh, func(deal *storagedeal.Deal) bool {
+		return deal.Miner == minerAddress
+	})
+
+	return outCh, nil
+}
+
+type dealLsPlumbing interface {
+	ConfigGet(string) (interface{}, error)
+	DealsIterator() (*query.Results, error)
+}
+
+// DealsLs returns an channel with all deals or a possible error
+func DealsLs(ctx context.Context, plumbing dealLsPlumbing) (<-chan *StorageDealLsResult, error) {
+	out := make(chan *StorageDealLsResult)
+	results, err := plumbing.DealsIterator()
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		defer close(out)
+		for entry := range (*results).Next() {
+			select {
+			case <-ctx.Done():
+				out <- &StorageDealLsResult{
+					Err: ctx.Err(),
+				}
+				return
+			default:
+				var storageDeal storagedeal.Deal
+				if err := cbor.DecodeInto(entry.Value, &storageDeal); err != nil {
+					out <- &StorageDealLsResult{
+						Err: errors.Wrap(err, "failed to unmarshal deals from datastore"),
+					}
+					return
+				}
+				out <- &StorageDealLsResult{
+					Deal: storageDeal,
+				}
+			}
+		}
+	}()
+
+	return out, nil
+}
+
+func filterDealChannel(dealCh <-chan *StorageDealLsResult, filterFunc func(*storagedeal.Deal) bool) <-chan *StorageDealLsResult {
+	outCh := make(chan *StorageDealLsResult)
+
+	go func() {
```

Comment:
This is still creating 2 goroutines where only 1 is needed (part of the point of moving the code out of plumbing). If you moved the DealLs code in here and have it also run filterFunc before emitting the value to the output channel,  then you could implement `DealClientLs`, `DealMinerLs` and `DealLs` by calling `filterDealChannel` with different filterFuncs (where DealLs's filterFunc always returns true).

Reply:
What's the advantage of using one goroutine instead of two? This seems more readable to me, and I don't see how it would have any impact on efficiency.

Reply:
It's less code and less system complexity. Less things to go wrong. For those reasons, I think it'd also be more readable. I'm not concerned about the efficiency. I'll leave it up to you, but I think its better not to use goroutines to pipe together channels where it can be avoided.